README.md
# NAME

Crypt::Password::StretchedHash - simple library for password hashing and stretching

# SYNOPSIS

This module provides Generation / Verification method for hashed password string.
There are two methods to handle parameters simply.

    use Test::More;
    use Crypt::Password::StretchedHash;
    use Digest::SHA;

    # crypt
    my $pwhash = Crypt::Password::StretchedHash->crypt(
        password        => q{password},
        hash            => Digest::SHA->new("sha256"),
        salt            => q{salt},
        stretch_count   => 5000,
        format          => q{base64},
    );
    is($pwhash, q{4hvvzqZio+l9vGifQ7xF2+FKiyWRcb4lV3OSo9PsfUw=});

    # verify
    my $result = Crypt::Password::StretchedHash->verify(
        password        => q{password},
        password_hash   => q{4hvvzqZio+l9vGifQ7xF2+FKiyWRcb4lV3OSo9PsfUw=},
        hash            => Digest::SHA->new("sha256"),
        salt            => q{salt},
        stretch_count   => 5000,
        format          => q{base64},
    );

    unless ( $result ) {
        # password error
    }

if you use class of the hash information(Crypt::Passwoed::SaltedHash::HashInfo),
there are two methods to generate/verify string for DB Store. 

    use Your::Password::HashInfo;
    use Crypt::Password::StretchedHash;
    

    my $hash_info = Your::Password::HashInfo->new;
    # crypt
    my $password = ...;
    my $pwhash_with_hashinfo = crypt_with_hashinfo(
        password    => $password,
        hash_info   => $hash_info,
    );
    

    # verify
    my $password = ...;
    my $pwhash_with_hashinfo = ...;
    my $result = verify_with_hashinfo(
        password        => $password,
        password_hash   => $pwhash_with_hashinfo,
        hash_info   => $hash_info,
    );
    

    unless ( $result ) {
        # password error
    }

# DESCRIPTION

Crypt::Password::StretchedHash is simple library for password hashing and stretching.
This module is available in generation and validation of the stretched password hash.

# METHODS

## crypt( %params ) : String

Generates stretched password hash.
This uses the following hash algorithm.

    for (1..$params{stretch_count}) {
        $hash->add( $pwhash, $params{password}, $salt );
        $pwhash = $hash->digest;
    }

- $params{password}

    This is password string.

- $params{hash}
 

    This is a hash function.
    This value must be the object of Digest::SHA or Digest::SHA3.

- $params{salt}
 

    This is salt string.

- $params{stretch\_count}

    This is stretching count.
    The value will depend on a kind of hash function and the server load.

- $params{format}

    This value is optional.
    If it has none, the password hash is returned with a binary string.
    If it has "hex", the password hash is returned with hexadecimal representation.
    If it has "base64", the password hash is returned with base64 representation.

## verify( %params ) : Int

Verifies stretched password hash.
This compares the value of $params{password\_hash} with the generated using crypt method.

## crypt\_with\_hashinfo( %params ) : String

Generates stretched password hash with hash information.

- $params{password}

    This is password string.

- $params{hash\_info}
 

    This is a hash information.
    You have to inherit [Crypt::Password::StretchedHash::HashInfo](http://search.cpan.org/perldoc?Crypt::Password::StretchedHash::HashInfo).

## verify\_with\_hashinfo( %params ) : Int

Verifies stretched password hash with hash information.
This compares the value of $params{password\_hash} with the generated using crypt method.

- $params{password}

    This is password string.

- $params{password\_hash}

    This is hashed password to be compared.

- $params{hash\_info}
 

    This is a hash information.
    You have to inherit [Crypt::Password::StretchedHash::HashInfo](http://search.cpan.org/perldoc?Crypt::Password::StretchedHash::HashInfo).

# LICENSE

Copyright (C) Ryo Ito.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Ryo Ito <ritou.06@gmail.com>


Changes
Revision history for Perl extension Crypt-Password-StretchedHash

{{$NEXT}}

0.05 2013-06-28T11:08:06Z

    - use Exporter
    - Fixed pod

0.04 2013-06-28T04:47:04Z

    - Added parent.pm to require modules #3

0.03 2013-06-24T13:28:27Z

    - Fixed pattern match with hash information

0.02 2013-06-24T06:29:11Z

    - Added HashInfo class and crypt/verify methods and removed OBJECT validation by Params::Validate

0.01 2013-06-19T04:28:57Z

    - original version



lib/Crypt/Password/StretchedHash/HashInfo.pm
package Crypt::Password::StretchedHash::HashInfo;

use strict;
use warnings;

sub new {
    my $class = shift;
    my $self = bless {}, $class;
    return $self;
}

sub delimiter {
    my $self = shift;
    die "This is abstract method. You have to return delimiter string.";
}

sub identifier {
    my $self = shift;
    die "This is abstract method. You have to return identifier of your HashInfo.";
}

sub hash {
    my $self = shift;
    die "This is abstract method. You have to return Digest::SHA or Digest::SHA3 object.";
}

sub salt {
    my $self = shift;
    die "This is abstract method. This is called at the time of first crypt. You have to return randomized salt string for each users.";
}

sub stretch_count {
    my $self = shift;
    die "This is abstract method. You have to return stretch count.";
}

sub format {
    my $self = shift;
    die "This is abstract method. You have to return format for hashed password representation.";
}

1;

__END__

=encoding utf-8

=head1 NAME

Crypt::Password::StretchedHash - Base class that specifies accessor for password information.

=head1 DESCRIPTION

Crypt::Password::StretchedHash::HashInfo is base class that specifies accessor for password information.
You have to inherit this, and implements subroutines according to the interface contract.

=head1 SYNOPSIS

You implement your HashInfo class as follows.

    package Your::Password::HashInfo;
    use parent 'Crypt::Password::StretchedHash::HashInfo';
    use Digest::SHA;
    use Crypt::OpenSSL::Random;
    use constant STRETCH_COUNT => 5000;
    
    sub delimiter {
        my $self = shift;
        return q{$};
    }
    
    sub identifier {
        my $self = shift;
        return q{1};
    }
    
    sub hash {
        my $self = shift;
        return Digest::SHA->new("sha256");
    }
    
    sub salt {
        my $self = shift;
        return Crypt::OpenSSL::Random::random_pseudo_bytes(32);
    }
    
    sub stretch_count {
        my $self = shift;
        return STRETCH_COUNT;
    }
   
    sub format {
        my $self = shift;
        return q{base64};
    }

By passing your hashinfo to Crypt::Password::StretchedHash->crypt_with_hashinfo method,
you obtain the hashed password with identifier and salt.

    use Crypt::Password::StretchedHash qw(
        crypt_with_hashinfo
    );
    use Your::Password::HashInfo;
    
    my $password = ...;
    my $hash_info = Your::Password::HashInfo->new;
    my $pwhash_with_hashinfo = crypt_with_hashinfo(
        password    => $password,
        hash_info   => $hash_info,
    );

It is similar at the time of the verification,
you pass your hashinfo to Crypt::Password::StretchedHash->verify_with_hashinfo method.

    use Crypt::Password::StretchedHash qw(
        verify_with_hashinfo
    );
    use Your::Password::HashInfo;
    
    my $password = ...;
    my $pwhash_with_hashinfo = ...;
    my $hash_info = Your::Password::HashInfo->new;
    my $is_valid = verify_with_hashinfo(
        password        => $password,
        password_hash   => $pwhash_with_hashinfo,
        hash_info   => $hash_info,
    );

=head1 METHODS

=head2 new : Object

constructor

=head2 delimiter : String

It returns delimiter string.
If delimiter is "$", generated string is as follows.

    $(identifier)$(salt)$(hashed password)

=head2 identifier : String

It returns identifier of hashinfo.
If delimiter is "$" and identifier is "1", generated string is as follows.

    $1$(salt)$(hashed password)

=head2 hash : Object

It returns hash object.
In the current version, only Digest::SHA and Digest::SHA3 are allowed.

=head2 stretch_count : Int

It returns stretching count, and if has to be an integer bigger than 0.

=head2 format : String

It returns hash object.
In the current version, only "hex" and "base64" are allowed.

=head2 salt : String

It returns salt string.
It may be binary strings.
If delimiter is "$" ,identifier is "1", format is "base64", salt is "test12345" generated string is as follows.

    $1$dGVzdDEyMzQ1$(hashed password)

=head1 LICENSE

Copyright (C) Ryo Ito.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Ryo Ito E<lt>ritou.06@gmail.comE<gt>

=cut



lib/Crypt/Password/StretchedHash.pm
package Crypt::Password::StretchedHash;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.04";

use Exporter qw(
    import
);
use Carp qw(
    croak
);
use MIME::Base64 qw(
    encode_base64
    decode_base64
);
use Params::Validate qw(
    SCALAR
);

our @EXPORT_OK = qw(
    crypt
    verify
    crypt_with_hashinfo
    verify_with_hashinfo
);

sub crypt {

    my %params = Params::Validate::validate(@_, {
        password        => { type => SCALAR },
        hash            => 1,
        salt            => { type => SCALAR },
        stretch_count   => { type => SCALAR, regex => qr/\A[0-9]+\z/,},
        format          => { type => SCALAR, optional => 1 },
    });

    my $salt = $params{salt};
    croak "\$params{hash} must be Digest::SHAx Object"
        unless ( $params{hash}->isa("Digest::SHA") || 
                $params{hash}->isa("Digest::SHA3"));

    my $hash = $params{hash};

    croak "\$params{stretch_count} must be more than 1"
        unless ($params{stretch_count} > 0);

    my $pwhash = q{};
    for (1..$params{stretch_count}) {
        $hash->add( $pwhash, $params{password}, $salt );
        $pwhash = $hash->digest;
    }

    if ( exists $params{format} && $params{format} eq q{hex} ){
        $pwhash = unpack("H*", $pwhash);
    }elsif( exists $params{format} && $params{format} eq q{base64} ){
        $pwhash = encode_base64 $pwhash;
        $pwhash =~ s/\n//;
        chomp($pwhash);
    }

    return $pwhash;
}

sub verify {

    my %params = Params::Validate::validate(@_, {
        password        => { type => SCALAR },
        password_hash   => { type => SCALAR },
        hash            => 1,
        salt            => { type => SCALAR },
        stretch_count   => { type => SCALAR, regex => qr/\A[0-9]+\z/,},
        format          => { type => SCALAR, optional => 1 },
    });

    my $pwhash = $params{password_hash};
    delete $params{password_hash};
    my $calculated_pwhash = Crypt::Password::StretchedHash::crypt( %params );
    return ( $calculated_pwhash eq $pwhash );
}

sub crypt_with_hashinfo {

    my %params = Params::Validate::validate(@_, {
        password    => { type => SCALAR },
        hash_info   => 1,
    });

    # validate hashinfo object
    my $hash_info = $params{hash_info};
    croak "\$params{hash_info} must be Crypt::Password::StretchedHash::HashInfo Object"
        unless ( $hash_info->isa("Crypt::Password::StretchedHash::HashInfo") );

    my $salt = $hash_info->salt;

    my $pwhash = Crypt::Password::StretchedHash::crypt(
        password        => $params{password},
        hash            => $hash_info->hash,
        salt            => $salt,
        stretch_count   => $hash_info->stretch_count,
        format          => $hash_info->format,
    );

    if ( $hash_info->format eq q{hex} ){
        $salt = unpack("H*", $salt);
    }elsif( $hash_info->format eq q{base64} ){
        $salt = encode_base64 $salt;
        chomp($salt);
    }
    
    return  $hash_info->delimiter . 
            $hash_info->identifier.
            $hash_info->delimiter.
            $salt.
            $hash_info->delimiter.
            $pwhash;

}

sub verify_with_hashinfo {

    my %params = Params::Validate::validate(@_, {
        password        => { type => SCALAR },
        password_hash   => { type => SCALAR },
        hash_info       => 1,
    });

    # validate hashinfo object
    my $hash_info = $params{hash_info};
    croak "\$params{hash_info} must be Crypt::Password::StretchedHash::HashInfo Object"
        unless ( $hash_info->isa("Crypt::Password::StretchedHash::HashInfo") );

    # split password hash
    my $delimiter = quotemeta($hash_info->delimiter);
    my $identifier = $hash_info->identifier;
    my ( $pwhash, $salt );

    my @split_password_hash = split( /$delimiter/, $params{password_hash} );
    if ($params{password_hash} =~ /\A$delimiter/ &&  
        scalar @split_password_hash == 4 &&
        $split_password_hash[0] eq q{} &&
        $identifier eq $split_password_hash[1] ){
        $salt = $split_password_hash[2];
        $pwhash = $split_password_hash[3];
    } else {
        return;
    }

    # obtain law_salt string
    if ( $hash_info->format eq q{hex} ){
        $salt = pack("H*", $salt);
    }elsif( $hash_info->format eq q{base64} ){
        $salt = decode_base64 $salt;
    }

    # generate hashed password
    my $expected_pwhash = Crypt::Password::StretchedHash::crypt(
        password        => $params{password},
        hash            => $hash_info->hash,
        salt            => $salt,
        stretch_count   => $hash_info->stretch_count,
        format          => $hash_info->format,
    );
    return ( $expected_pwhash eq $pwhash );

}

1;
__END__

=encoding utf-8

=head1 NAME

Crypt::Password::StretchedHash - simple library for password hashing and stretching

=head1 SYNOPSIS

This module provides Generation / Verification method for hashed password string.
There are two methods to handle parameters simply.

    use Crypt::Password::StretchedHash qw(
        crypt
        verify
    );
    use Digest::SHA;

    # crypt
    my $pwhash = crypt(
        password        => q{password},
        hash            => Digest::SHA->new("sha256"),
        salt            => q{salt},
        stretch_count   => 5000,
        format          => q{base64},
    );

    # verify
    my $result = verify(
        password        => q{password},
        password_hash   => q{4hvvzqZio+l9vGifQ7xF2+FKiyWRcb4lV3OSo9PsfUw=},
        hash            => Digest::SHA->new("sha256"),
        salt            => q{salt},
        stretch_count   => 5000,
        format          => q{base64},
    );


if you use class of the hash information(Crypt::Passwoed::SaltedHash::HashInfo),
there are two methods to generate/verify string for DB Store. 

    use Your::Password::HashInfo;
    use Crypt::Password::StretchedHash qw(
        crypt_with_hashinfo
        verify_with_hashinfo
    );
    
    my $hash_info = Your::Password::HashInfo->new;
    # crypt
    my $password = ...;
    my $pwhash_with_hashinfo = crypt_with_hashinfo(
        password    => $password,
        hash_info   => $hash_info,
    );
    
    # verify
    my $password = ...;
    my $pwhash_with_hashinfo = ...;
    my $result = verify_with_hashinfo(
        password        => $password,
        password_hash   => $pwhash_with_hashinfo,
        hash_info   => $hash_info,
    );

=head1 DESCRIPTION

Crypt::Password::StretchedHash is simple library for password hashing and stretching.
This module is available in generation and validation of the stretched password hash.

=head1 METHODS

=head2 crypt( %params ) : String

Generates stretched password hash.
This uses the following hash algorithm.

    for (1..$params{stretch_count}) {
        $hash->add( $pwhash, $params{password}, $salt );
        $pwhash = $hash->digest;
    }

=over
 
=item $params{password}

This is password string.

=item $params{hash}
 
This is a hash function.
This value must be the object of Digest::SHA or Digest::SHA3.

=item $params{salt}
 
This is salt string.

=item $params{stretch_count}

This is stretching count.
The value will depend on a kind of hash function and the server load.

=item $params{format}

This value is optional.
If it has none, the password hash is returned with a binary string.
If it has "hex", the password hash is returned with hexadecimal representation.
If it has "base64", the password hash is returned with base64 representation.

=back

=head2 verify( %params ) : Int

Verifies stretched password hash.
This compares the value of $params{password_hash} with the generated using crypt method.

=head2 crypt_with_hashinfo( %params ) : String

Generates stretched password hash with hash information.

=over

=item $params{password}

This is password string.

=item $params{hash_info}
 
This is a hash information.
You have to inherit L<Crypt::Password::StretchedHash::HashInfo>.

=back

=head2 verify_with_hashinfo( %params ) : Int

Verifies stretched password hash with hash information.
This compares the value of $params{password_hash} with the generated using crypt method.

=over

=item $params{password}

This is password string.

=item $params{password_hash}

This is hashed password to be compared.

=item $params{hash_info}
 
This is a hash information.
You have to inherit L<Crypt::Password::StretchedHash::HashInfo>.

=back

=head1 LICENSE

Copyright (C) Ryo Ito.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Ryo Ito E<lt>ritou.06@gmail.comE<gt>

=cut

